app/models/puppetclass.rb

```diff
@@ -87,31 +84,25 @@ def klass
     name.gsub(module_name+"::","")
   end
 
-  # return host ids from config groups by type
-  def host_ids_from_config_groups(host_type)
-    ids = config_groups.joins(:host_config_groups)
-                 .where("host_config_groups.host_type='#{host_type}'")
-                 .pluck('host_config_groups.host_id') unless config_group_classes.empty?
-    ids || []
-  end
-
   def all_hostgroups(with_descendants = true)
-    ids = hostgroup_ids
-    ids += host_ids_from_config_groups('Hostgroup')
-    hgs = Hostgroup.unscoped.where(:id => ids.uniq)
+    hgs = Hostgroup.authorized
+                   .reorder('')
+                   .eager_load(:hostgroup_classes, :config_groups => [:config_group_classes])
+                   .where("#{self.id} IN (hostgroup_classes.puppetclass_id, config_group_classes.puppetclass_id)")
+                   .uniq
```

Comment:
added the parameter, `all_hostgroups` is also used in the ui. 
`host_type` is now automatically applied by the `eager_load`, which rails now knows how to do as we start the query from the hostgroup rather then the config group as it was before.